Ship v6 fixes to prod (#67)
iExecBlockchainComputing/documentation: src/guides/build-iapp/advanced/result-callback.md

---

> title: Result Callback Guide

Reviewer: ```suggestion
title: Result callback guide
```

description:
  Use the iExec result callback to have the protocol invoke a function on your

>   smart contract at the end of a (optionally confidential TEE) task execution.

Reviewer: ```suggestion
  Use the iExec result callback feature to have the protocol invoke a function on your
  smart contract at the end of a task execution.
```

---

> # Result Callback

Reviewer: ```suggestion
# Result callback
```

Reviewer: And all other titles.

Reviewer 2: fix Commit: 240e37f9c160a44b8c3f2e8d76d035bcd3b34734

This guide explains how to trigger a callback function at the end of a
successful task on your smart contract.

Use a callback when your smart contract should:

- Ingest off-chain computed data (API aggregation, ML inference, analytics) and
  persist it
- React to an execution outcome (conditional trigger, state transition)
- Store a timestamped record (price feed, score, KPI, proof hash)
- Bridge logic between external systems and on-chain state

> ## 🧩 High-Level Flow

Reviewer: ```suggestion
## 🧩 High-level flow
```

> 1. A requester executes an iApp.

Reviewer: ```suggestion
1. A requester deploys the smart contract that should receive the callback data.
2. The requester executes an iApp and specifies the callback address.
```

2. The iApp writes `${IEXEC_OUT}/computed.json` with a `callback-data` field
   (ABI‑encoded bytes you crafted).
3. After the task completes and is validated, the iExec protocol invokes your
   contract’s `receiveResult(bytes32,bytes)`.
4. Your contract decodes and processes those bytes if callback data have been
   provided.

## Step-by-Step Implementation

> ### Step 1: Prepare the Callback Payload in the iApp

Reviewer: I would invert step 1 and step 2 because it makes more sense to deploy the callback contract first.

Reviewer 2: yeah ! Fix Commit: aec3b168844334d2cd48b4d737b1db5f9ad5c3fc

> You only need to write `computed.json` containing the key `callback-data`.  

Reviewer: `callback-data` should be equal to `resutlDigest`.
https://github.com/iExecBlockchainComputing/PoCo/blob/00a111e7ba35cc9a61796eb65a06936b32078db2/contracts/facets/IexecPoco2Facet.sol#L209

Reviewer 2: As we’ve seen, the resultDigest is automatically calculated by the worker.
No need for us to expose this complexity to our builders

That value must be the ABI‑encoded bytes your contract knows how to decode.  
Example tuple schema we'll use:
`(uint256 timestamp, string pairAndPrecision, uint256 scaledValue)`.

```ts twoslash
import { writeFileSync } from 'node:fs';
import { AbiCoder } from 'ethers';

const timestamp = Math.floor(Date.now() / 1000);
const pair = 'BTC-USD';
const scaled = '9';

> // ---cut---

Reviewer: What does it mean?

Reviewer 2: Well, twoslash tests/verify that the code works correctly. Using `// ---cut---` allows you to hide pieces of code (like hard-coded values) that are only there to prevent TypeScript errors. I decided to add that, because I don’t think those hard-coded values brings any real value for the builder.

async function main() {
  // Your business logic here ...

  const abiCoder = new AbiCoder();
  const abiPayload = abiCoder.encode(
    ['uint256', 'string', 'uint256'],
    [timestamp, pair, scaled]
  );

  writeFileSync(
    `${process.env.IEXEC_OUT}/computed.json`,
    JSON.stringify({
      'callback-data': abiPayload,
    })
  );
}
```

### Step 2: Implement the Callback Contract

> Your contract must expose `receiveResult(bytes32,bytes)`

Reviewer: ```suggestion
Your contract must expose the function `receiveResult(bytes32,bytes)`
```

[ERC1154](https://github.com/iExecBlockchainComputing/iexec-solidity/blob/master/contracts/ERC1154/IERC1154.sol).
The protocol calls it with:

> - `_callID`: the first arg the taskId

Reviewer: ```suggestion
- `_callID`: This parameter represents the `taskId`, passed as the first argument
```
